File: backend/app/services/dialogue/gpt_client.py
import asyncio
import time
import math
from typing import Dict, List, Optional, Any, TypedDict, AsyncIterator
from datetime import datetime
import aiohttp
import json
import logging
from pyee import EventEmitter

logger = logging.getLogger(__name__)

class GPTConfig(TypedDict, total=False):
    """Configuration for GPT API requests."""
    model: str
    temperature: Optional[float]
    maxTokens: Optional[int]
    stop: Optional[List[str]]

class GPTResponse(TypedDict, total=False):
    """Response from GPT API with generated text and usage statistics."""
    text: str
    usage: Optional[Dict[str, int]]
    raw: Optional[Any]
    error: Optional[str]

class GPTClientOptions(TypedDict, total=False):
    """Configuration options for GPTClient."""
    apiKey: str
    apiUrl: Optional[str]
    rateLimit: Optional[int]
    timeoutMs: Optional[int]
    maxRetries: Optional[int]
    backoffBaseMs: Optional[int]

class GPTUsageStats(TypedDict):
    """Usage statistics for token consumption."""
    totalTokens: int
    rollingTokens: int
    windowMs: int
    lastReset: int

class GPTClient(EventEmitter):
    """
    GPTClient: Handles GPT/OpenAI API requests, prompt/context management, 
    and emotion context formatting for LLMs.
    Emits events for request, error, and fallback. Provides static helpers for prompt injection.
    """
    
    def __init__(self, options: GPTClientOptions):
        """Initialize the GPTClient with API configuration."""
        super().__init__()
        self.api_key = options['apiKey']
        self.api_url = options.get('apiUrl', 'https://api.openai.com/v1/chat/completions')
        self.rate_limit = options.get('rateLimit', 60)
        self.timeout_ms = options.get('timeoutMs', 10000)
        self.max_retries = options.get('maxRetries', 3)
        self.backoff_base_ms = options.get('backoffBaseMs', 500)
        self.requests = []  # Timestamps of recent requests for rate limiting
        self.usage_window_ms = 60000  # 1 minute window
        self.usage_stats = {
            'totalTokens': 0,
            'rollingTokens': 0,
            'windowMs': self.usage_window_ms,
            'lastReset': int(time.time() * 1000)
        }
        
        # Headers for API requests
        self.headers = {
            'Authorization': f'Bearer {self.api_key}',
            'Content-Type': 'application/json'
        }
    
    async def generate_completion(self, prompt: str, context: List[str], config: GPTConfig) -> GPTResponse:
        """
        Generates a GPT completion given a prompt and context.
        Emits 'request', 'error', and 'fallback' events.
        """
        await self._enforce_rate_limit()
        messages = [
            {'role': 'system', 'content': c} for c in context
        ]
        messages.append({'role': 'user', 'content': prompt})
        
        # Prepare API payload
        payload = {
            **config,
            'messages': messages,
            'temperature': config.get('temperature', 0.7),
            'max_tokens': config.get('maxTokens', 512),
        }
        if 'stop' in config and config['stop']:
            payload['stop'] = config['stop']
            
        self.emit('request', {'prompt': prompt, 'context': context, 'config': payload})
        
        # Attempt API request with retry logic
        try:
            async with aiohttp.ClientSession() as session:
                async with session.post(
                    self.api_url,
                    headers=self.headers,
                    json=payload,
                    timeout=self.timeout_ms / 1000  # aiohttp uses seconds
                ) as response:
                    if response.status != 200:
                        error_text = await response.text()
                        raise aiohttp.ClientResponseError(
                            request_info=response.request_info,
                            history=response.history,
                            status=response.status,
                            message=error_text,
                            headers=response.headers
                        )
                    
                    data = await response.json()
                    choice = data.get('choices', [{}])[0].get('message', {}).get('content', '')
                    if 'usage' in data:
                        self._update_usage(data['usage'].get('total_tokens'))
                    
                    return {
                        'text': choice,
                        'usage': data.get('usage'),
                        'raw': data
                    }
        except Exception as error:
            self.emit('error', error)
            logger.error(f"GPT API error: {str(error)}")
            
            # Retry logic with exponential backoff
            for attempt in range(1, self.max_retries + 1):
                try:
                    await asyncio.sleep(2 ** attempt * self.backoff_base_ms / 1000)
                    async with aiohttp.ClientSession() as session:
                        async with session.post(
                            self.api_url,
                            headers=self.headers,
                            json=payload,
                            timeout=self.timeout_ms / 1000
                        ) as response:
                            if response.status != 200:
                                continue
                                
                            data = await response.json()
                            choice = data.get('choices', [{}])[0].get('message', {}).get('content', '')
                            if 'usage' in data:
                                self._update_usage(data['usage'].get('total_tokens'))
                            
                            return {
                                'text': choice,
                                'usage': data.get('usage'),
                                'raw': data
                            }
                except Exception as retry_error:
                    if attempt == self.max_retries:
                        self.emit('fallback', retry_error)
                        logger.error(f"All retry attempts failed: {str(retry_error)}")
                        return {'text': '', 'error': str(retry_error)}
            
            return {'text': '', 'error': str(error)}
    
    async def _enforce_rate_limit(self):
        """Enforces the rate limit by delaying if necessary."""
        now = int(time.time() * 1000)
        # Filter out requests older than 1 minute
        self.requests = [t for t in self.requests if now - t < 60000]
        
        if len(self.requests) >= self.rate_limit:
            # Wait until we're allowed to make another request
            wait_time = 60000 - (now - self.requests[0])
            await asyncio.sleep(wait_time / 1000)  # Convert to seconds
        
        self.requests.append(int(time.time() * 1000))
    
    def _update_usage(self, tokens: Optional[int]):
        """Updates token usage statistics."""
        now = int(time.time() * 1000)
        if now - self.usage_stats['lastReset'] > self.usage_window_ms:
            self.usage_stats['rollingTokens'] = 0
            self.usage_stats['lastReset'] = now
        
        if tokens:
            self.usage_stats['totalTokens'] += tokens
            self.usage_stats['rollingTokens'] += tokens
    
    def get_usage_stats(self) -> GPTUsageStats:
        """Returns current usage statistics."""
        return dict(self.usage_stats)
    
    @staticmethod
    async def sleep(ms: int):
        """Utility to sleep for ms milliseconds."""
        await asyncio.sleep(ms / 1000)
    
    @staticmethod
    def count_tokens(text: str) -> int:
        """Counts tokens in a string (approximate, for OpenAI models)."""
        # Simple approximation: 1 token ≈ 4 characters (for English)
        return math.ceil(len(text) / 4)
    
    @staticmethod
    def format_emotion_context_for_prompt(emotions: List[Dict[str, Any]]) -> str:
        """
        Formats emotion context for prompt injection.
        @param emotions Array of emotion objects with name, intensity, and optional description
        @returns String for prompt injection
        """
        if not emotions:
            return 'No emotions are currently active.'
        
        formatted_emotions = []
        for emotion in emotions:
            emotion_str = f"{emotion['name']} (intensity: {emotion['intensity']}"
            if 'description' in emotion and emotion['description']:
                emotion_str += f", {emotion['description']}"
            emotion_str += ")"
            formatted_emotions.append(emotion_str)
        
        return 'Current emotions: ' + ', '.join(formatted_emotions) 

File: Visual_DM/Visual_DM/Assets/Scripts/Dialogue/GPTClient.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using UnityEngine;
using UnityEngine.Networking;
using System.Text;
using VisualDM.Core;

namespace VisualDM.Dialogue
{
    /// <summary>
    /// Client for interacting with GPT/OpenAI API
    /// </summary>
    public class GPTClient : MonoBehaviour
    {
        [Header("API Configuration")]
        [SerializeField] private string apiUrl = "https://api.openai.com/v1/chat/completions";
        [SerializeField] private int requestTimeoutSeconds = 10;
        [SerializeField] private int maxRetries = 3;
        [SerializeField] private int backoffBaseMs = 500;

        [Header("Rate Limiting")]
        [SerializeField] private int rateLimit = 60; // Max requests per minute
        [SerializeField] private int windowMs = 60000; // 60 seconds

        private string _apiKey;
        private List<long> _requests = new List<long>();
        private GPTUsageStats _usageStats;
        private EventBus _eventBus;

        [Serializable]
        private class GPTUsageStats
        {
            public int totalTokens;
            public int rollingTokens;
            public int windowMs;
            public long lastReset;
        }

        [Serializable]
        private class ChatMessage
        {
            public string role;
            public string content;
        }

        [Serializable]
        private class ChatRequestBody
        {
            public string model;
            public List<ChatMessage> messages;
            public float temperature;
            public int max_tokens;
            public List<string> stop;
        }

        [Serializable]
        private class ChatResponseBody
        {
            public ChatChoice[] choices;
            public ChatUsage usage;

            [Serializable]
            public class ChatChoice
            {
                public ChatMessage message;
                public int index;
                public string finish_reason;
            }

            [Serializable]
            public class ChatUsage
            {
                public int prompt_tokens;
                public int completion_tokens;
                public int total_tokens;
            }
        }

        private void Awake()
        {
            _usageStats = new GPTUsageStats
            {
                totalTokens = 0,
                rollingTokens = 0,
                windowMs = windowMs,
                lastReset = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
            };
            
            // Get the EventBus from the scene
            _eventBus = EventBus.Instance;
            
            // Try to get the API key from PlayerPrefs or a secure storage
            _apiKey = PlayerPrefs.GetString("OpenAI_APIKey", "");
        }

        /// <summary>
        /// Sets the API key for OpenAI requests
        /// </summary>
        public void SetApiKey(string apiKey)
        {
            _apiKey = apiKey;
            PlayerPrefs.SetString("OpenAI_APIKey", apiKey);
        }

        /// <summary>
        /// Generates a completion from GPT given a prompt and context
        /// </summary>
        public async Task<GPTResponse> GenerateCompletionAsync(string prompt, List<string> context, GPTConfig config)
        {
            // Check if API key is set
            if (string.IsNullOrEmpty(_apiKey))
            {
                return new GPTResponse
                {
                    text = "",
                    error = "API key not set",
                    raw = null
                };
            }
            
            await EnforceRateLimitAsync();
            
            // Create the messages array from context and prompt
            var messages = new List<ChatMessage>();
            foreach (var ctx in context)
            {
                messages.Add(new ChatMessage { role = "system", content = ctx });
            }
            messages.Add(new ChatMessage { role = "user", content = prompt });
            
            // Create the request body
            var requestBody = new ChatRequestBody
            {
                model = config.model,
                messages = messages,
                temperature = config.temperature,
                max_tokens = config.maxTokens,
                stop = config.stop
            };
            
            // Emit request event
            _eventBus.Emit("gpt_request", new Dictionary<string, object>
            {
                { "prompt", prompt },
                { "context", context },
                { "config", config }
            });
            
            // Send the request
            try
            {
                return await SendRequestWithRetriesAsync(requestBody);
            }
            catch (Exception ex)
            {
                _eventBus.Emit("gpt_error", ex);
                return new GPTResponse
                {
                    text = "",
                    error = ex.Message,
                    raw = ex
                };
            }
        }

        /// <summary>
        /// Sends the GPT request with retry logic
        /// </summary>
        private async Task<GPTResponse> SendRequestWithRetriesAsync(ChatRequestBody requestBody)
        {
            Exception lastException = null;
            
            // Try initial request
            try
            {
                return await SendRequestAsync(requestBody);
            }
            catch (Exception ex)
            {
                lastException = ex;
                _eventBus.Emit("gpt_error", ex);
            }
            
            // Retry with exponential backoff
            for (int attempt = 1; attempt <= maxRetries; attempt++)
            {
                try
                {
                    // Wait with exponential backoff
                    int delayMs = (int)Math.Pow(2, attempt) * backoffBaseMs;
                    await Task.Delay(delayMs);
                    
                    return await SendRequestAsync(requestBody);
                }
                catch (Exception ex)
                {
                    lastException = ex;
                    
                    if (attempt == maxRetries)
                    {
                        _eventBus.Emit("gpt_fallback", ex);
                    }
                }
            }
            
            // All attempts failed
            return new GPTResponse
            {
                text = "",
                error = lastException?.Message ?? "Unknown error during API request",
                raw = lastException
            };
        }

        /// <summary>
        /// Sends a single request to the GPT API
        /// </summary>
        private async Task<GPTResponse> SendRequestAsync(ChatRequestBody requestBody)
        {
            // Convert request body to JSON
            string jsonBody = JsonUtility.ToJson(requestBody);
            
            // Create web request
            using (UnityWebRequest request = new UnityWebRequest(apiUrl, "POST"))
            {
                byte[] bodyRaw = Encoding.UTF8.GetBytes(jsonBody);
                request.uploadHandler = new UploadHandlerRaw(bodyRaw);
                request.downloadHandler = new DownloadHandlerBuffer();
                request.SetRequestHeader("Content-Type", "application/json");
                request.SetRequestHeader("Authorization", $"Bearer {_apiKey}");
                request.timeout = requestTimeoutSeconds;
                
                // Send request
                var operation = request.SendWebRequest();
                
                // Wait for completion
                while (!operation.isDone)
                {
                    await Task.Yield();
                }
                
                // Check for errors
                if (request.result != UnityWebRequest.Result.Success)
                {
                    throw new Exception($"API Error: {request.responseCode} - {request.error} - {request.downloadHandler.text}");
                }
                
                // Parse response
                string responseJson = request.downloadHandler.text;
                ChatResponseBody responseBody = JsonUtility.FromJson<ChatResponseBody>(responseJson);
                
                if (responseBody == null || responseBody.choices == null || responseBody.choices.Length == 0 || 
                    responseBody.choices[0].message == null)
                {
                    throw new Exception("Invalid API response format");
                }
                
                // Update token usage
                if (responseBody.usage != null)
                {
                    UpdateUsage(responseBody.usage.total_tokens);
                }
                
                // Return response
                return new GPTResponse
                {
                    text = responseBody.choices[0].message.content,
                    usage = new GPTResponse.GPTUsage
                    {
                        promptTokens = responseBody.usage?.prompt_tokens ?? 0,
                        completionTokens = responseBody.usage?.completion_tokens ?? 0,
                        totalTokens = responseBody.usage?.total_tokens ?? 0
                    },
                    raw = responseBody,
                    error = null
                };
            }
        }

        /// <summary>
        /// Enforces the rate limit by waiting if necessary
        /// </summary>
        private async Task EnforceRateLimitAsync()
        {
            long now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            
            // Remove timestamps older than window
            _requests.RemoveAll(t => now - t >= windowMs);
            
            // If at rate limit, wait until we can make a new request
            if (_requests.Count >= rateLimit && _requests.Count > 0)
            {
                long wait = windowMs - (now - _requests[0]);
                if (wait > 0)
                {
                    await Task.Delay((int)wait);
                }
            }
            
            // Add current request timestamp
            _requests.Add(DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
        }

        /// <summary>
        /// Updates token usage statistics
        /// </summary>
        private void UpdateUsage(int tokens)
        {
            long now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            
            // Reset rolling tokens if window has passed
            if (now - _usageStats.lastReset > _usageStats.windowMs)
            {
                _usageStats.rollingTokens = 0;
                _usageStats.lastReset = now;
            }
            
            // Update token counts
            if (tokens > 0)
            {
                _usageStats.totalTokens += tokens;
                _usageStats.rollingTokens += tokens;
            }
        }

        /// <summary>
        /// Returns current usage statistics
        /// </summary>
        public GPTUsageStats GetUsageStats()
        {
            return _usageStats;
        }

        /// <summary>
        /// Formats emotion context for prompt injection
        /// </summary>
        public static string FormatEmotionContextForPrompt(List<EmotionState> emotions)
        {
            if (emotions == null || emotions.Count == 0)
            {
                return "No emotions are currently active.";
            }
            
            List<string> formattedEmotions = new List<string>();
            foreach (var emotion in emotions)
            {
                string desc = !string.IsNullOrEmpty(emotion.description) ? $", {emotion.description}" : "";
                formattedEmotions.Add($"{emotion.name} (intensity: {emotion.intensity}{desc})");
            }
            
            return "Current emotions: " + string.Join(", ", formattedEmotions);
        }

        /// <summary>
        /// Emotion state for prompt injection
        /// </summary>
        [Serializable]
        public class EmotionState
        {
            public string name;
            public float intensity;
            public string description;
        }
        
        /// <summary>
        /// Estimates the token count in a text string
        /// </summary>
        public static int CountTokens(string text)
        {
            // Simple approximation: 1 token ≈ 4 characters (for English)
            if (string.IsNullOrEmpty(text)) return 0;
            return Mathf.CeilToInt(text.Length / 4f);
        }
    }
} 